src/notebooks/deepnote/environmentCapture.node.ts

```diff
@@ -0,0 +1,248 @@
+import { inject, injectable } from 'inversify';
+import { execFile } from 'node:child_process';
+import os from 'node:os';
+import { promisify } from 'node:util';
+
+import type { Environment } from '@deepnote/blocks';
+
+import { PythonEnvironment } from '../../platform/pythonEnvironments/info';
+import { computeHash } from '../../platform/common/crypto';
+import { logger } from '../../platform/logging';
+import { parsePipFreezeFile } from './pipFileParser';
+import { IDeepnoteEnvironmentManager, IDeepnoteNotebookEnvironmentMapper } from '../../kernels/deepnote/types';
+import { Uri } from 'vscode';
+import { DeepnoteEnvironment } from '../../kernels/deepnote/environments/deepnoteEnvironment';
+import * as path from '../../platform/vscode-path/path';
+
+const captureTimeoutInMilliseconds = 5_000;
+
+/**
+ * Wraps a promise with a timeout. If the promise doesn't resolve within the timeout,
+ * returns the fallback value instead of throwing.
+ */
+async function withTimeout<T>(promise: Promise<T>, timeoutMs: number, fallback: T, operationName: string): Promise<T> {
+    let timeoutId: NodeJS.Timeout | undefined;
+
+    const timeoutPromise = new Promise<T>((resolve) => {
+        timeoutId = setTimeout(() => {
+            logger.warn(`[EnvironmentCapture] ${operationName} timed out after ${timeoutMs}ms`);
+            resolve(fallback);
+        }, timeoutMs);
+    });
+
+    try {
+        const result = await Promise.race([promise, timeoutPromise]);
+
+        return result;
+    } finally {
+        if (timeoutId) {
+            clearTimeout(timeoutId);
+        }
+    }
+}
```

Comment:
nit: might be worth extracting to a shared helper? Have you also checked in the codebase if `raceTimeout` is doing something similar?